src/lib.rs

```diff
@@ -227,6 +236,9 @@ pub struct Peripherals {
     pub gpt2: hal::gpt::Unclocked,
     /// DMA channels
     pub dma: hal::dma::Unclocked,
+    /// The SysTick delay timer
+    #[cfg(feature = "systick")]
+    pub systick: SysTick,
```

Comment:
Having feature-gated fields on structs can break stuff when adding their associated features due to rust's exhaustive pattern matching - I think this is normally avoided by keeping the fields private to disallow exhaustively matching on the struct and exposing them via methods instead. In our case however I think this is probably totally fine, especially as exposing only some peripherals via methods would inevitably lead to ownership issues that would be much more annoying to deal with heh!

Reply:
Great call. It sounds like we need to prevent exhaustive pattern matching, and a [`#[non_exhaustive]` attribute](https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute) on `Peripherals` might be perfect for that. That should let users enable features without introducing compilation errors.

Users who were using features, but want to disable features, would still need to guard against these members disappearing, just as they would have to deal with conditionally-compiled methods disappearing.